always-hydrate.md

```diff
@@ -0,0 +1,133 @@
+prefetch --folders` commands to bulk download the files they care about. And they can further specify
+the `--hydrate` flag to either command to bulk hydrate those files into the working directory.
+
+However, this is not a great experience because it's just a one-time prefetch and hydrate. What that
+means is that the next time the projection changes, e.g. after a `git checkout`, anywhere from some
+to all of those files can become dehydrated, and the user will again pay the cost of file-by-file
+downloads while trying to build.
+
+With this feature, we want the developer to make one gesture and communicate that a set of files
+should always be kept hydrated, even after the projection changes. That being said, a secondary goal
+is to not add any significant time to commands like `git checkout` that change the projection, so the
+implementation should ensure that these are done as background downloads that do not block these
+commands.
+
+## UX
+This is really an extension of the existing prefetch functionality. As such, it makes sense to tie
+the UX to the existing `gvfs prefetch` verb.
+
+Side note on the existing prefetch verb: it's really two different commands in one, and it's
+unfortunate that they are combined. The command `gvfs prefetch --commits` is entirely different in
+its goals, usage, and implementation than either of `gvfs prefetch --files` or `gvfs prefetch 
+--folders` (and the latter two can be combined with each other, but not with `--commits`). Ideally we
+would have had `gvfs prefetch-commits` and `gvfs prefetch-files` as two separate commands, which
+would have then led to a cleaner design for this new addition. I'm open to other suggestions here
+if anyone sees a better pattern for extending the current verbs.
+
+### Proposed UX for always-hydrate commands:
+Adding new entries: 
+
+Add an `--always-hydrate` flag to the existing `--files` and `--folders` variants of `gvfs prefetch`.
+This will communicate that whatever list of folders, files, and wildcards that were passed in to the
+command should be saved and reapplied after every projection change
+
+Syntax:
+
+```
+    gvfs prefetch 
+      [<Repo Root Path>]
+      [--files <semicolon-delimited list of file paths, prefix * wildcards allowed>] 
+      [--folders <semicolon-delimited list of folder paths, no wildcards>]
+      [--folders-list <path of file containing newline-delimited paths of folders, no wildcards>] 
+      [--[always-]hydrate]
+```
+
+The only change from today's syntax is the optional `[always-]` on the hydrate flag. 
+`--always-hydrate` implies `--hydrate`.
+
+Passing in the `--always-hydrate` flag has all the same behaviors that `prefetch --hydrate` has today
+and also adds the specified files and folders to the `always-hydrate` file, so that those files will
+continue to be hydrated in the future.
+
+Managing existing entries:
+
+In addition to adding entries to the `always-hydrate` file, users will also want to list out the
```

Comment:
> I think that's exactly what we're doing :-)

I can see it both ways :)

My thinking had been along these lines:

- `prefetch` -> synchronous VFS for Git command that will download objects and\or hydrate files.

- `always-hydrate` -> VFS for Git feature that enables more aggressive downloading of objects and hydrating of files.

But maybe that's over thinking it.  Users need to run `gvfs mount` to enable auto-mount, and that's similar to the experience here.

Reply:
To answer your question though - no I don't think `gvfs config` would be an appropriate place to add this because that verb is design to capture machine-wide config, not repo-specific. E.g. what upgrade ring you want and whether to automount.

I do think that the `--always-hydrate` flag makes good sense on the prefetch verb. It's actually doing the prefetching and hydrating, but _also_ recording that we want to keep prefetching and hydrating these paths later. 

The other two flags for listing and removing entries definitely feel strange here, and maybe that's what you meant. Those two variants don't do any prefetching, rather they just read or remove from the `always-hydrate` file. I'm open to ideas on where they should go, but overall my feeling was that it makes more sense for them to be on the same verb as the `--always-hydrate` flag, and that one makes the most sense on the existing `prefetch` verb.

Reply:
I think it would be helpful to users if we follow what git has done with the excludes and have a machine-wide and repo specific always-hydrate.  Most users I think would want to set the machine-wide list and then just have it work, whether I clone, prefetch, checkout, etc.  If I am creating a new clone, having to find the right repo, get the list of files or folders, copy it, clone, prefetch with the list seems like a lot of steps.

Reply:
I don't agree that most users will want that functionality. And I don't think we envision that users will have to do much of this manually. The Office team is building up their own tool that analyzes their build dependencies, figures out what components the dev will need, and prefetch those files. We're giving them the ability to additionally indicate that they don't want to just prefetch once, but also keep this files/folders hydrated in the future.

That being said, the ability to configure some machine-wide defaults would be an easy extension of this design. I think we could easily add it in later on, but it's not something that I think we should include in an initial release. Also keep in mind that we have no other such pattern of machine-wide and repo-local config settings for VFS for Git, and doing so adds lots of complexities that I'd rather avoid if customers aren't actually asking for it.

Reply:
What if we use the gvfs section of the git config to store this information?  That would give us the machine-wide vs repo setting for free and we wouldn't have another file to manage.  Listing, adding, deleting would all be taken care of as well.  I don't know how many files and folders we plan on saving here so that could be a limitation?

Reply:
You might have a folder named Foo in two completely unrelated repos, and you wouldn't want your global config affecting them both. So we'd have to tie the config to a URL at least, though even that is unreliable.